fuzzers/main.py

```diff
@@ -0,0 +1,949 @@
+
+    bucket = gcs_client.bucket(GCS_BUCKET)
+
+    repo = Repository(REPO_DIR, REPO_URL)
+    repo.clone_if_need_be()
+
+    corpus = Corpus(CORPUS_DIR, bucket)
+
+    while True:
+        if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+            return
+
+        sync_log_files = []
+        date = datetime.datetime.utcnow().strftime('%Y-%m-%d')
+
+        # Read the configuration from the repository
+        cfg = repo.latest_config('master')
+        cfg_for = {
+            b['name']: repo.latest_config(b['name']) for b in cfg['branch']
+        }
+
+        # Figure out which targets we want to run
+        branches = random_weighted(cfg['branch'], NUM_FUZZERS, "branches")
+        targets = [
+            random_weighted(cfg_for[b['name']]['target'], 1, "target")[0]
+            for b in branches
+        ]
+
+        # Synchronize the relevant corpuses
+        corpus.stop_synchronizing()
+        corpus.update()
+        for targ in set((t['crate'], t['runner']) for t in targets):
+            log_path = pathlib.Path('sync') / date / targ[0] / targ[1] / str(
+                uuid.uuid4())
+            utils.mkdirs((LOGS_DIR / log_path).parent)
+            # pylint: disable=consider-using-with
+            corpus.synchronize(targ[0], targ[1],
+                               open(LOGS_DIR / log_path, 'a', encoding='utf-8'))
+            # pylint: enable=consider-using-with
+            sync_log_files.append(log_path)
+            if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+                return
+
+        # Initialize the fuzzers
+        fuzzers = []
+        for (branch, target) in zip(branches, targets):
+            worktree = repo.worktree(branch['name'])
+            log_path = pathlib.Path(
+                'fuzz') / date / target['crate'] / target['runner'] / str(
+                    uuid.uuid4())
+            log_file = LOGS_DIR / log_path
+            utils.mkdirs(log_file.parent)
+            fuzzers.append(
+                FuzzProcess(corpus_vers=corpus.version,
+                            branch=branch,
+                            target=target,
+                            repo_dir=worktree,
+                            log_relpath=log_path,
+                            log_fullpath=log_file))
+
+        # Build the fuzzers
+        for fuzzer in fuzzers:
+            fuzzer.build()
+            if pause_exit_spot(pause_evt, resume_evt, exit_evt):
+                return
+
+        # Start the fuzzers
+        atexit.register(kill_fuzzers, bucket, fuzzers)
+        for fuzzer in fuzzers:
+            fuzzer.start(corpus)
+
+        # Wait until something happens
+        started = time.monotonic()
+        last_sync_file_upload = started
+        next_restart = started + AUTO_REFRESH_INTERVAL.total_seconds()
+        while time.monotonic() < next_restart:
+            # Avoid busy-looping by sleeping 100ms between each loop run
+            time.sleep(0.1)
```

Comment:
We definitely could, but then it'd take one more second to actually pause the fuzzer, which given the average time of running tests that I'm incidentally noticing when looking at logs (without any statistics made), which is in the range of a few minutes usually, I'm not sure it'd still be reactive enough? And to the best of my knowledge the loop should be pretty efficient anyway so it shouldn't slow down fuzzing much.

That said if you think 1s of delay would be ok I'd be happy to do it :)

(Also, incidentally the more I work on this the more I think I should have gone with rust instead of python, because then I could have made it be async with sum-type-based messaging channels between actors, which would both have avoided the need for this, and allowed for other easy improvements like the one about measuring build time… I had forgotten how useful sum types were as my recent years had been rust and they're natural there :sweat_smile: )

Reply:
The tests won’t usually do anything crucial during the first second so I feel it should be fine.  I don’t really have strong opinions here.

Dunno if it’s really about sum type (after all being dynamically typed Python kinda by definition has sum types and typing.Union is a thing) but rather about being able to wait on multiple events which threading.Event isn’t great at.

Reply:
Got it, going with 1 second :) also for sum types you're right, what I was meaning was more along the lines of easy matches than sum types per se